### PR TITLE
Int 8576 cron gets stuck if report scores cannot be updated

### PR DESCRIPTION
**Pull request by hperperidis**

### CODE REVIEW

File: lang/en/plagiarism_turnitin.php [hunk 1]
```diff
@@ -224,3 +224,4 @@
 $string['privacy:metadata:plagiarism_turnitin_client:submission_filename'] = 'The name of the submitted file is sent to Turntin so that it is identifiable.';
 
 $string['privacy:metadata:core_files'] = 'Turnitin Assignment stores files that have been uploaded to Moodle to form a Turnitin submission.';
+$string['errorcode13'] = 'This submissionid - {$a->externalid} was not found in TII. Unable to retrieve similarity score and other submission data.';
```
Comment (dwinn) [inline lang/en/plagiarism_turnitin.php]: I'd change TII to Turnitin.

File: lib.php [hunk 2]
```diff
@@ -2032,6 +2060,55 @@ public function cron_update_scores($submissiontype = 'file', $submissions) {
         return true;
     }
 
+    private function checklocalsubmissionstate($assignmentIds, $submissionids) {
```
Comment (dwinn) [inline lib.php]: Methods can have underscores in Moodle so I'd change this to `change_local_submission_state()` for readability.

File: lib.php [hunk 3]
```diff
@@ -2032,6 +2060,55 @@ public function cron_update_scores($submissiontype = 'file', $submissions) {
...
+        // Initialise Comms Object.
+        $turnitincomms = new turnitin_comms();
+        $turnitincall = $turnitincomms->initialise_api();
+        $tiisubmissionids = array();
+
+        foreach ($assignmentIds as $assignmentId) {
+            $submission = new TiiSubmission();
+            $submission->setAssignmentId($assignmentId);
+            $submission->setDateFrom('2018-09-12T09:00:00Z'); // TODO(hperperidis): see how we can get a valid date here;
```
Comment (dwinn) [inline lib.php]: This looks like a work in progress. I think hardcoding the date like this could cause issues.

File: lib.php [hunk 4]
```diff
@@ -2032,6 +2060,55 @@ public function cron_update_scores($submissiontype = 'file', $submissions) {
...
+            }
+        }
+
+        return array(
+            'trimmedSubmissions' => array_intersect($submissionids, $tiisubmissionids),
+            'missingTiiSubmissions' => array_diff($submissionids, $tiisubmissionids)
+        );
+    }
+
+    private function invalidatemissingsubmission($missingsubmission) {
```
Comment (dwinn) [inline lib.php]: Similarly a better name would be `invalidate_missing_submission`.

File: lib.php [hunk 5]
```diff
@@ -2032,6 +2060,55 @@ public function cron_update_scores($submissiontype = 'file', $submissions) {
...
+            'id, 
+            cm, 
+            externalid, 
+            userid'
+        );
+        $plagiarismfile = new stdClass();
+        $plagiarismfile->id = $currentsubmission->id;
+        $plagiarismfile->externalid = $currentsubmission->externalid;
+        $plagiarismfile->userid = $currentsubmission->userid;
+        $plagiarismfile->statuscode='error';
```
Comment (dwinn) [inline lib.php]: Some spacing needed around the `=`

File: lib.php [hunk 6]
```diff
@@ -1952,14 +1954,41 @@ public function cron_update_scores($submissiontype = 'file', $submissions) {
...
+                    if (!isset(array_flip($assignmentids)[$moduleturnitinconfig->value])) {
+                        $assignmentids[] = $moduleturnitinconfig->value;
+                    }
                 }
             }
         }
 
-        if (count($submissionids) > 0) {
+        //Seperate out $submissionids that do not exist in TII.
+        $validatedSubmissions = $this->checklocalsubmissionstate($assignmentids, $submissionids);
```
Comment (dwinn) [inline lib.php]: Agree with Clegg, we can't use variables with camel case as they fail Travis check. Variables should be all lowercase and no underscores. Not something I personally agree with but we have to follow the Moodle standards.

File: lib.php [hunk 7]
```diff
@@ -2032,6 +2060,55 @@ public function cron_update_scores($submissiontype = 'file', $submissions) {
         return true;
     }
 
+    private function checklocalsubmissionstate($assignmentIds, $submissionids) {
+        // Initialise Comms Object.
+        $turnitincomms = new turnitin_comms();
+        $turnitincall = $turnitincomms->initialise_api();
+        $tiisubmissionids = array();
+
+        foreach ($assignmentIds as $assignmentId) {
```
Comment (dwinn) [inline lib.php]: Camels

File: lang/en/plagiarism_turnitin.php (hunk 1 above)
Comment (hperperidis) [reply lang/en/plagiarism_turnitin.php]: 👍

File: lib.php (hunk 6 above)
Comment (hperperidis) [reply lib.php]: 👍

File: lib.php [hunk 8]
```diff
@@ -1970,10 +1970,10 @@ public function cron_update_scores($submissiontype = 'file', $submissions) {
             }
         }
 
-        //Seperate out $submissionids that do not exist in TII.
+        // Seperate out $submissionids that do not exist in TII.
```
Comment (dwinn) [inline lib.php]: `Separate`

File: lib.php [hunk 9]
```diff
@@ -2060,21 +2060,21 @@ public function cron_update_scores($submissiontype = 'file', $submissions) {
         return true;
     }
 
-    private function check_local_submission_state($assignmentIds, $submissionids) {
+    private function check_local_submission_state($assignmentids, $submissionids) {
```
Comment (dwinn) [inline lib.php]: Need to rename this function to match the call to it without underscores.

**Pull request merged**